### DISCUSSION TITLE
Complied/Standalone EXE for Windows

### DISCUSSION THREAD

**Discussion by philhemel** [avantrec/soco-cli#56] [General]
Hi folks, is there a Complied/Standalone EXE for Windows or can anyone help me make one?
I've got the script working fine on a test PC but I'm not allowed to install Python on the PC I'd really like to use it on.
I've tried using pyinstaller which almost works but fails with errors when running it!
Many thanks

**Comment by pwt** [comment]
Hi Phil: I've no plans to create standalone executable versions, partly because I've no experience of doing so and partly because I don't want to complicate the build/distribution process (especially if multi-platform support is required).

Others may be able to assist, and I'd be happy to post instructions.

**Reply by philhemel** [reply]
Ok thanks

**Comment by philhemel** [comment]
UPDATE - I've managed to make a working standalone EXE file as follows in case it helps anyone else:

1.
Edit "sonos.py" in Notepad
(should be in C:\Users\Whoever\AppData\Local\Programs\Python\Python311\Lib\site-packages\soco_cli)

2.
CHANGE THE FOLLOWING LINE:
from .wait_actions import process_wait
TO:
from soco_cli.wait_actions import process_wait

This fixes error in the standalone exe:
	Traceback (most recent call last):
	  File "soco_cli\sonos.py", line 36, in <module>
	ImportError: attempted relative import with no known parent package
	[1840] Failed to execute script 'sonos' due to unhandled exception!

3.
ADD THE FOLLOWING LINE (I added it immedately below the line above):
from sys import exit

This fixes error in the standalone exe:
	Traceback (most recent call last):
	  File "soco_cli\sonos.py", line 514, in <module>
	NameError: name 'exit' is not defined
	[9252] Failed to execute script 'sonos' due to unhandled exception!

4.
Save changes and close notepad

5.
Run CMD as Administrator, then run the following from it:

python -m pip install pyinstaller

CD "C:\Users\Whoever\AppData\Local\Programs\Python\Python311\Lib\site-packages\soco_cli"

pyinstaller sonos.py --onefile

6.
Standalone "sonos.exe" should then be in folder "C:\Users\Whoever\AppData\Local\Programs\Python\Python311\Lib\site-packages\soco_cli\dist"
I've copied this to a PC that has no Python files installed at all, and it's  working a treat :)

**Reply by pwt** [reply]
Thanks!

I'll check to see if the changes you had to make are required more generally, and include them in the standard distribution if so.

**Reply by philhemel** [reply]
Thank you!
